pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -1504,14 +1546,89 @@ func isDeviceOpened(deviceToDetach AttachedVolume, mounter mount.Interface) (boo
 	return deviceOpened, nil
 }
 
-func isMigrated(vpm *volume.VolumePluginMgr, spec *volume.Spec) bool {
+func isNodeMigratable(og *operationGenerator, pluginName string, nodeName types.NodeName) (bool, error) {
+	// TODO(dyzz): Write a test that runs with feature flags on, tries to do an inline volume with a migratable volume but without
+	// driver installed. Should get the correct error message. Then clean up. Try to do a migratable volume but WITH driver installed,
+	// should succeed (even better if we can check that the driver did something).
+	var err error
+	var nodeInfo *csiapi.CSINodeInfo
+
+	pm, err := og.volumePluginMgr.IsPluginMigratableByName(pluginName)
+	if err != nil {
+		return false, err
+	}
+
+	if !pm {
+		// Feature flags aren't on so migration as a whole is just turned off
+		return false, nil
+	}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSINodeInfo) {
+		return false, fmt.Errorf("failed to check if node migrated, CSINodeInfo feature gate not enabled")
+	}
+
+	if og.csiNodeInfoHasSynced() {
+		nodeInfo, err = og.csiNodeInfoLister.Get(string(nodeName))
+		if err != nil {
+			return false, fmt.Errorf("failed to get CSI node info for node %v from informer: %v", string(nodeName), err)
+		}
+	} else {
+		glog.Warningf("CSINodeInfo informer not synced, please check that CSINodeInfo CRD is installed. If so, this warning should not appear again after ~1 minute")
```

Comment:
I am a little confused about the race you described. Are you saying that if informer is synced it may have stale API objects? I was under the impression that synced informer means that any `GET` to the informer will guarantee to be the freshest object, otherwise what's the point in having an inconsistent cache?

The only two options I see here in the current code are:
1) Informer synced  -> get from informer (freshest object)
2) Informer not yet synced -> get from API Server (freshest object)

Reply:
Yes, informers can return old data. It does List + Watch and the watch can have delay on network. That's OK. Informer will let you know when it gets new data.

GET from API server can get old data too - at the time your client parses the data there may be already new on the server. This time, you don't get event when it changes.

In this particular case, GET from API server could be faster than initial List + Watch and return something newer than the informer. Under "normal" conditions the informer syncs very quickly, so this discussion is more or less academical.